### ISSUE TITLE
Cannot use `mbedtls_x509_crt_ext_cb_t` (certificate extension callback) with `mbedtls_ssl_conf_verify()`

### ISSUE THREAD

**Issue by dimakuv** [Mbed-TLS/mbedtls#7182]
### Suggested enhancement

The callback `mbedtls_x509_crt_ext_cb_t` allows to examine unknown-to-mbedTLS x.509 certificate extensions, and allow the certificate if the callback returns non-negative value.

This allows to implement a callback required for the new OID being standardized in the [TCG DICE specification](https://members.trustedcomputinggroup.org/wg/DICE/document/36858). The OID is `2.23.133.5.4.9` and is called `tcg-dice-conceptual-message-wrapper` (previously called `tcg-dice-tagged-evidence`). This OID is marked a **critical extension**.

mbed TLS by default **fails on unknown critical extensions**. So to verify a certificate with this new OID, we must use the above callback.

Unfortunately, it seems that invoking this callback is only implemented for the `mbedtls_x509_crt_parse_der_with_ext_cb()` function; this function is found in the `libmbedx509.so` library.

However, for our implementation in the [Interoperable RA-TLS](https://github.com/CCC-Attestation/interoperable-ra-tls/blob/main/docs/Interoperable-RA-TLS-cert-edvidence-formats.md) and [Gramine](https://github.com/gramineproject/gramine/pull/1039) projects, we would need a similar function that works at the level of mbedTLS's TLS library `libmbedtls.so`.

There seems to be nothing of the sort in `libmbedtls.so`. E.g., here is the excerpt of the exported symbols:
```
$ nm --defined-only libmbedtls.so | grep verify
00000000000163e0 T mbedtls_ssl_conf_verify
00000000000177f0 T mbedtls_ssl_get_verify_result
000000000001a960 T mbedtls_ssl_set_calc_verify_md
0000000000016560 T mbedtls_ssl_set_verify
0000000000015760 t ssl_calc_verify_tls_sha256
00000000000154d0 t ssl_calc_verify_tls_sha384
```

Currently, Gramine [uses](https://github.com/gramineproject/gramine/blob/b741bd113fddac4a5c04902056638b34f88b5f8b/CI-Examples/ra-tls-mbedtls/src/client.c#L344) `mbedtls_ssl_conf_verify()` with the callback, that allows to correctly verify **non-critical certificate extensions**. However, as soon as we enable the "criticality" flag in the extension, mbedTLS's TLS handshake fails even before jumping into the callback.

**Summary**: the callback to verify certificate extensions `mbedtls_x509_crt_ext_cb_t` is implemented for X.509 parsing code (part of `libmbedx509.so`), but is not implemented for TLS handshake code (part of `libmbedtls.so`).
- Is there a plan to implement similar callbacks for the TLS handshake logic?
- Is there any work around for this problem?

The tested mbed TLS version is v3.3.0.

### Justification

Mbed TLS needs this because there seems to be no way of using `libmbedtls.so` library with **currently-unknown-to-mbedTLS certificate extensions**.

### Additional materials

Please see discussions:
- https://github.com/CCC-Attestation/interoperable-ra-tls/issues/9
- https://github.com/CCC-Attestation/interoperable-ra-tls/pull/11

Related mbedTLS issues:
- https://github.com/Mbed-TLS/mbedtls/issues/3241
- https://github.com/Mbed-TLS/mbedtls/issues/4378

**Comment by gilles-peskine-arm** [comment]
I'm labeling this as a bug because while it requires a new feature (a new `ssl_conf` function), it's technically a regression in Mbed TLS 3.0, since in 2.x there was the (clumsy) workaround of compiling with `MBEDTLS_X509_ALLOW_UNSUPPORTED_CRITICAL_EXTENSION`.

**Comment by askourtis** [comment]
This is breaking for us as well, is there any update on the status? Thank you!

**Comment by gilles-peskine-arm** [comment]
Unfortunately it's unlikely that we'll have time to work on this in the short or medium term.

Would you be willing to submit a patch that works for you? I can't make a firm promise, but that would greatly increase the chances that this gets resolved.

The patch would need:

* to not break the existing API (feel free to propose something for a design review);
* to have unit tests.

**Comment by ReeceSX** [comment]
> Unfortunately it's unlikely that we'll have time to work on this in the short or medium term.

Including this? https://github.com/Mbed-TLS/mbedtls/issues/7079

So we're confirming a TLS library intended for client devices, not really for serving major servers, hasn't had viable pinning for years and wont have viable certificate pinning for the "medium term"? 

You would rather waste time getting your formatting and hard-coded past expired certs to pass established tests, than be able to verify a peers accurately?

What even are mbedtls's priories anymore? As far as I'm concerned, all those past notices about mbedtls not supporting 1.3 (a 2018 spec) are still valid so long as version negation is still up in their air, peers are completely unverifiable past one hard-coded certificate, and any future proofing in the form of quic interfaces are open to question.

**Comment by gilles-peskine-arm** [comment]
> What even are mbedtls's priories anymore?

Mbed TLS's priority is resource-constrained devices that often don't use “general” CAs. This has been the case ever since Mbed TLS was called Mbed TLS. This ecosystem doesn't depend critically on certificate pinning.

> a TLS library intended for client devices, not really for serving major servers

I wouldn't exactly say that. A lot of embedded servers use Mbed TLS. Major servers aren't our priority target, but neither are desktop/laptop/smartphone clients.

> peers are completely unverifiable past one hard-coded certificate

??? Mbed TLS does verify certificate chains. Do you mean that you want to check the revocation status of certificates on every connection? When your ecosystem doesn't have hundreds of root CAs and has a very tight communication budget, pervasive revocation checks are rarely an acceptable cost-benefit compromise.

> any future proofing in the form of quic interfaces are open to question

Again, we're a crypto and TLS library, not a QUIC library. QUIC is relevant to the web but less so in IoT and so far it isn't in our priority list.

**Comment by ReeceSX** [comment]
> This ecosystem doesn't depend critically on certificate pinning

tl;dr "this tls library doesn't depend on certificate pinning"

I'm sorry, but how do you not see where you're going wrong? This library cannot reasonable perform as either a client while it cannot pin certificates arbitrarily or as a server while negation is up in the air. 

> When your ecosystem doesn't have hundreds of root CAs and has a very tight communication budget, pervasive revocation checks are rarely an acceptable cost-benefit compromise.

And what if we didn't want to use your abstraction of certificate authorities? You know, like how CURL has CURLOPT_SSL_CTX_FUNCTION, Java SSE's X509TrustManager, C#'s ServicePointManager.ServerCertificateValidationCallback, etc.

> Again, we're a crypto and TLS library, not a QUIC library. QUIC is relevant to the web but less so in IoT and so far it isn't in our priority list.

You acknowledge the callback problem is a bug; you acknowledge it's a regression in 1.3; you can surely acknowledge there is a demand in the form of issues from godot users, this user, et al.

I don't care about QUIC support. I care about a TLS library not fully supporting a 2018 specification, while also not being able to arbitrarily pin certificates as a client-centeric library.

**Comment by mpg** [comment]
> > Unfortunately it's unlikely that we'll have time to work on this in the short or medium term.
> 
> Including this? #7079
> 
@ReeceSX  If you're unhappy that some issues aren't fixed yet, and want to let us know they're blocking you, please comment on the relevant issues (and create new ones if necessary) rather than derailing other discussions. I'd also suggest sticking to technical points relevant to the issue at hand; overly general rants seem less likely to be constructive. Thank you.
